desc/coils.py

```diff
@@ -6,16 +6,107 @@
 
 import numpy as np
 
-from desc.backend import jnp
-from desc.compute import rpz2xyz, xyz2rpz_vec
+from desc.backend import jnp, tree_stack, tree_unstack, vmap
+from desc.compute import get_params, rpz2xyz, xyz2rpz_vec
 from desc.geometry import (
     FourierPlanarCurve,
     FourierRZCurve,
     FourierXYZCurve,
     SplineXYZCurve,
 )
-from desc.magnetic_fields import _MagneticField, biot_savart
-from desc.utils import flatten_list
+from desc.grid import LinearGrid
+from desc.magnetic_fields import _MagneticField
+from desc.utils import equals, errorif, flatten_list
+
+
+def biot_savart_hh(eval_pts, coil_pts_start, coil_pts_end, current):
+    """Biot-Savart law for filamentary coils following [1].
+
+    The coil is approximated by a series of straight line segments
+    and an analytic expression is used to evaluate the field from each
+    segment.
+
+    Parameters
+    ----------
+    eval_pts : array-like shape(n,3)
+        Evaluation points in cartesian coordinates
+    coil_pts_start, coil_pts_end : array-like shape(m,3)
+        Points in cartesian space defining the start and end of each segment.
+        Should be a closed curve, such that coil_pts_start[0] == coil_pts_end[-1]
+        though this is not checked.
+    current : float
+        Current through the coil (in Amps).
+
+    Returns
+    -------
+    B : ndarray, shape(n,3)
+        magnetic field in cartesian components at specified points
+
+    [1] Hanson & Hirshman, "Compact expressions for the Biot-Savart
+    fields of a filamentary segment" (2002)
+    """
+    d_vec = coil_pts_end - coil_pts_start
+    L = jnp.linalg.norm(d_vec, axis=-1)
+
+    Ri_vec = eval_pts[jnp.newaxis, :] - coil_pts_start[:, jnp.newaxis, :]
+    Ri = jnp.linalg.norm(Ri_vec, axis=-1)
+    Rf = jnp.linalg.norm(
+        eval_pts[jnp.newaxis, :] - coil_pts_end[:, jnp.newaxis, :], axis=-1
+    )
+    Ri_p_Rf = Ri + Rf
+
+    # 1.0e-7 == mu_0/(4 pi)
```

Comment:
Is this just more numerically stable? If you are going to hard-code numbers, you might as well combine it with the factor of 2 and use `2.0e-7`

Reply:
yeah its just to save some flops